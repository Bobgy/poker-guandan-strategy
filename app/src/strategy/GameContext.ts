import {
  NaturalRankWithoutJokers,
  NaturalRank,
  PowerRank,
  RED_JOKER,
  BLACK_JOKER,
  getUsualPowerRank,
} from './utils'

export class GameContext {
  constructor(mainRank: NaturalRankWithoutJokers) {
    this.mainRank = mainRank
  }
  /** Which rank is largest in current game play. */
  readonly mainRank: NaturalRankWithoutJokers
  /**
   * Returns range[0, 14], power of the rank in this gameplay.
   * The weakest is 0. The strongest (red joker) is 14.
   */
  getOrder(rank: NaturalRank): PowerRank {
    if (rank == RED_JOKER) {
      return 14
    } else if (rank == BLACK_JOKER) {
      return 13
    } else if (rank == this.mainRank) {
      return 12
    } else {
      const usualRank = getUsualPowerRank(rank)
      const mainRankUsualRank = getUsualPowerRank(this.mainRank)
      if (usualRank > mainRankUsualRank) {
        return (usualRank - 1) as PowerRank
      } else {
        return usualRank as PowerRank
      }
    }
  }
}
