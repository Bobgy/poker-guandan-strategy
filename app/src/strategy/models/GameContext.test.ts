import {
  A,
  NATURAL_RANKS_WITHOUT_JOKERS,
  NATURAL_RANKS,
  NATURAL_RANK,
} from './const'

import { GameContext } from './GameContext'

describe('GameContext', () => {
  it('keeps a mainRank', () => {
    const context = new GameContext(A)
    expect(context.mainRank).toEqual(A)
  })
  NATURAL_RANKS_WITHOUT_JOKERS.map(mainRank => {
    describe(`When main rank is ${mainRank}`, () => {
      it('understands card order', () => {
        const context = new GameContext(mainRank)
        const ranks = NATURAL_RANKS.map(rank => ({
          label: NATURAL_RANK[rank].label,
          rank,
          order: context.getOrder(rank),
        }))
        ranks.sort((a, b) => a.order - b.order)
        expect(ranks.map(rank => rank.label).join(' ')).toMatchSnapshot()
        const orderSet = new Set(ranks.map(rank => rank.order))
        // sanity check for order set
        expect(orderSet.size).toEqual(15)
        orderSet.forEach(order => expect(order >= 0 && order < 15))
      })
    })
  })
})
