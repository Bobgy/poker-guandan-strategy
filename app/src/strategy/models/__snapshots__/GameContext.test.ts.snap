// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameContext When main rank is 1 understands card order 1`] = `"2 3 4 5 6 7 8 9 10 J Q K A BJ RJ"`;

exports[`GameContext When main rank is 2 understands card order 1`] = `"3 4 5 6 7 8 9 10 J Q K A 2 BJ RJ"`;

exports[`GameContext When main rank is 3 understands card order 1`] = `"2 4 5 6 7 8 9 10 J Q K A 3 BJ RJ"`;

exports[`GameContext When main rank is 4 understands card order 1`] = `"2 3 5 6 7 8 9 10 J Q K A 4 BJ RJ"`;

exports[`GameContext When main rank is 5 understands card order 1`] = `"2 3 4 6 7 8 9 10 J Q K A 5 BJ RJ"`;

exports[`GameContext When main rank is 6 understands card order 1`] = `"2 3 4 5 7 8 9 10 J Q K A 6 BJ RJ"`;

exports[`GameContext When main rank is 7 understands card order 1`] = `"2 3 4 5 6 8 9 10 J Q K A 7 BJ RJ"`;

exports[`GameContext When main rank is 8 understands card order 1`] = `"2 3 4 5 6 7 9 10 J Q K A 8 BJ RJ"`;

exports[`GameContext When main rank is 9 understands card order 1`] = `"2 3 4 5 6 7 8 10 J Q K A 9 BJ RJ"`;

exports[`GameContext When main rank is 10 understands card order 1`] = `"2 3 4 5 6 7 8 9 J Q K A 10 BJ RJ"`;

exports[`GameContext When main rank is 11 understands card order 1`] = `"2 3 4 5 6 7 8 9 10 Q K A J BJ RJ"`;

exports[`GameContext When main rank is 12 understands card order 1`] = `"2 3 4 5 6 7 8 9 10 J K A Q BJ RJ"`;

exports[`GameContext When main rank is 13 understands card order 1`] = `"2 3 4 5 6 7 8 9 10 J Q A K BJ RJ"`;
